cider/exec/nextgen/jitlib/base/JITValue.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef JITLIB_BASE_JITVALUE_H
+#define JITLIB_BASE_JITVALUE_H
+
+#include <memory>
+
+#include "exec/nextgen/jitlib/base/ValueTypes.h"
+
+namespace jitlib {
+class JITValue;
+class JITFunction;
+
+class JITValuePointer {
+ public:
+  JITValuePointer(std::nullptr_t ptr) : ptr_(ptr) {}
+
+  template <typename T>
+  JITValuePointer(std::unique_ptr<T>&& ptr) : ptr_(ptr.release()) {}
+
+  JITValue* get() { return ptr_.get(); }
+
+ public:
+  JITValue& operator*() { return *ptr_; }
+
+  JITValue* operator->() { return ptr_.get(); }
+
+  operator JITValue&() { return *ptr_; }
+
+ private:
+  std::unique_ptr<JITValue> ptr_;
+};
+
+class JITValue {
+ public:
+  explicit JITValue(JITTypeTag type_tag,
+                    JITFunction& parent_function,
+                    const std::string& name,
+                    JITBackendTag backend)
+      : value_name_(name)
+      , parent_function_(parent_function)
+      , type_tag_(type_tag)
+      , backend_tag_(backend) {}
+
+  virtual ~JITValue() = default;
```

Comment:
can JITValue be designed not using virtual table?

Reply:
JITValue was designed as virtual inheritance because of:
- We need to keep the possibility for using different JIT backend, so the JITValue and JITFunction should be implemented as virtual inheritance or template.
    - The template, could be optimized at compilation time completely with better performance. However, due to the template tends to spread to the user, the user must implement some template interfaces in header files and destroy the encapsulation. For example, if we design JITValue as ```JITValue<Backend>```, the user ```translator.consume()``` should be designed as ```translator.consume<Backend>()``` and it must be defined at the header.
    - Virtual inheritance avoids this problem and provides better interfaces to the user. The shortage of virtual inheritance is the performance, the costs come from vtable lookup and virtual function can't be inlined so complete optimization at compilation time is forbidden. Actually, cost of vtable lookup is very trivial in modern compiler based on some reports. Meanwhile, the benefit of inlining functions in JITValue may not significant because the logic in the functions is a little complicated.  Therefore, for more user-friend usage of JITLib, JITValue was designed as virtual inheritance.

Reply:
OK, and all inheritance class should be defined as final derived class.